invenio_access/permissions.py

```diff
@@ -117,64 +118,78 @@ def _cache_key(action_need):
         """Helper method to generate cache key."""
         return get_action_cache_key(
             action_need.value,
-            action_need.argument if hasattr(action_need, 'argument') else None
+            action_need.argument if hasattr(action_need, "argument") else None,
         )
 
+    @staticmethod
+    def _split_actionsneeds(needs):
+        """Split needs into sets of ActionNeed and any other *Need."""
+        action_needs, other_needs = set(), set()
+        for need in needs:
+            action_needs.add(need) if need.method == "action" else \
+                other_needs.add(need)
+        return action_needs, other_needs
+
     def _load_permissions(self):
-        """Load permissions associated to actions."""
+        """Load permissions for all needs expanding actions."""
         result = _P(needs=set(), excludes=set())
-        if not self.allow_by_default:
-            result.needs.update(self.explicit_needs)
-            result.excludes.update(self.explicit_excludes)
-
-        for explicit_need in self.explicit_needs:
-            if explicit_need.method == 'action':
-                # in-place update of results
-                result.update(self._load_action(explicit_need))
-            elif self.allow_by_default:
-                result.needs.add(explicit_need)
-
-        for explicit_exclude in self.explicit_excludes:
-            if explicit_exclude.method == 'action':
-                # in-place update of results
-                result.update(self._load_action(explicit_exclude))
-            elif self.allow_by_default:
-                result.excludes.add(explicit_exclude)
+
+        # split ActionNeeds and any other *Needs in separates Sets
```

Comment:
Minor: *Needs\*

Reply:
actually, here I wanted to say `UserNeed`, `RoleNeed` or any other `*Need`, that's why... but it is clear that it is not clear! :)